### ISSUE TITLE
IDS or IPS function

### ISSUE THREAD

**Issue by Threefish4096** [AdguardTeam/AdGuardHome#2206]
Is it possible to integrate in AdGuard Home, combining the functions of the DHCP server, the functions of an Intrusion Detection System or a Prevention System?
Perhaps by providing the ability to deep packet inspection on packets, inspecting network traffic or other methods.
Many thanks in advance!

**Comment by ameshkov** [comment]
Any examples of such tools?

**Comment by Threefish4096** [comment]
For example, I use a device called BananaPi R2, on it I have AdGuard Home installed.
I already use the features you offer in the context of DHCP to assign the IP to each device on my LAN.
The BananaPi R2 is connected to a Fritz Box 7530, now .. If I set the BananaPi as an exposed device, similar to DMZ, I would get that all the traffic goes through it both when it goes on the Internet and when it comes back from the Internet, the requests already pass us of DNS ..

The wonderful thing would be if it is possible to integrate the possibility of analyz all the packets that pass through my device, BananaPi, through AdGuard and with deep packets inspection see if any packets bring something unwanted.

Exploits, virus and more could be kept up to date through some claud platform designed like the current blacklists you use to block ads (GitHub or GitLab link or other platforms), maybe you can integrate the fingerprint of new threats by acquiring them via APIs from services that publish them and which are security milestones computer technology.(https://www.exploit-db.com/  , https://nvd.nist.gov/vuln/data-feeds , https://www.cvedetails.com/)

Something like this drawing, forgive my handwriting (https://prnt.sc/v3tibr)
I tried to explain what my network is like and what would be the features it would be nice to have, I put Firewall not by chance lol

These are some features that would be very nice to have:
https://prnt.sc/v3st80
https://prnt.sc/v3su3n
https://prnt.sc/v3suf1

These are some existing software that does the functionality of IPS:
https://prnt.sc/v3tlr1

Thank you so much for your attention and if you need anything else, don't hesitate to tag me!

**Comment by ameshkov** [comment]
So basically you'd like us to become the gateway and non only filter DNS queries, but also perform DPI.

And once the DPI part is done, there are tons of features that can be built upon this.

This would effectively transform AdGuard Home into a router. Well, that'd be great and that's why I'll keep this feature request open. But this will only be possible in a quite distant future.

**Comment by Threefish4096** [comment]
I will look forward to that future with open arms 💯 

P.S I also have other little things that could be added to the whole ;)

**Comment by Onepamopa** [comment]
For DPI - nDPI can be used, based on DPDK and actively maintained.

**Comment by lordraiden** [comment]
This doesn't make any sense,  maybe in 10 years but if you want firewall you can already use opnsense

**Comment by KizunaH** [comment]
https://www.ntop.org/products/deep-packet-inspection/ndpi/
https://github.com/ntop/nDPI
